contrib/win32/win32compat/misc.c

```diff
@@ -1044,58 +1044,62 @@ w32_strerror(int errnum)
 		return _sys_errlist_ext[errnum - EADDRINUSE];
 	return strerror(errnum);
 }
-/* 
- * Temporary implementation of readpassphrase. 
- * TODO - this needs to be reimplemented as per 
- * https://linux.die.net/man/3/readpassphrase
- */
-char * 
-readpassphrase(const char *prompt, char *out, size_t out_len, int flags) {
-	char *askpass = NULL;
-	char *ret = NULL;
 
-	DWORD mode;
-	size_t len = 0;
-	int retr = 0;
+char *
+readpassphrase(const char *prompt, char *outBuf, size_t outBufLen, int flags) {
+	int currentIndex = 0;
+	char ch;
+
+	if (outBufLen == 0) {
+		errno = EINVAL;
+		return(NULL);
+	}
+	
+	while (_kbhit()) _getch();
 
-	/* prompt user */
 	wchar_t* wtmp = utf8_to_utf16(prompt);
 	if (wtmp == NULL)
 		fatal("unable to alloc memory");
+
 	_cputws(wtmp);
 	free(wtmp);
 
-	len = retr = 0;
-
-	while (_kbhit())
-		_getch();
-
-	while (len < out_len) {
-		out[len] = (unsigned char)_getch();
-
-		if (out[len] == '\r') {
-			if (_kbhit()) /* read linefeed if its there */
-				_getch();
+	while (currentIndex < outBufLen - 1) {
+		ch = _getch();
+		
+		if (ch == '\r') {
+			if (_kbhit()) _getch(); /* read linefeed if its there */				
 			break;
-		}
-		else if (out[len] == '\n') {
+		} else if (ch == '\n') {
 			break;
-		}
-		else if (out[len] == '\b') { /* backspace */
-			if (len > 0)
-				len--; /* overwrite last character */
-		}
-		else if (out[len] == '\003') {
-			/* exit on Ctrl+C */
+		} else if (ch == '\b') { /* backspace */
+			if (currentIndex > 0) {
+				if (flags & RPP_ECHO_ON)
+					printf("%c \b", ch);
+
+				currentIndex--; /* overwrite last character */
+			}
+		} else if (ch == '\003') { /* exit on Ctrl+C */
```

Comment:
I am not familar with key stroke. but some characters like ESC, the logic will fall to this line printf("%c", ch); what will be displayed?